### ISSUE TITLE
Implement `<ViewOnlyMenu />`

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#4812]
## Feature Description

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* A new `ViewOnlyMenu` component should be implemented, following the design in [this Figma file](https://www.figma.com/file/i04x83uA9U8HHBANlhS7Db/Dashboard-Sharing?node-id=1512%3A17482).
* As indicated by the full Figma design, this menu will eventually replace the `UserMenu` in the view-only version of the dashboard. However, as part of this issue that shouldn't happen yet. This issue is purely about implementing the component and relying on Storybook for preview.
* To list the services which are being shared with the current user, the component should rely on the `getViewableModules()` selector from #4805.
* If the current user has the capability to authenticate with Site Kit (`GOOGLESITEKIT_AUTHENTICATE`), the UI should be slightly expanded as seen [here in the Figma file](https://www.figma.com/file/i04x83uA9U8HHBANlhS7Db/Dashboard-Sharing?node-id=784%3A12652):
    * The text at the top is a bit different and indicates that the user can sign in, and right below it a sign in with Google button should be placed, which should take the user right to the setup flow (i.e. the same logic should be triggered that happens when you click "Sign in with Google" on the splash screen, with the user eventually landing on the OAuth consent screen via the proxy).
    * Each shared module should be accompanied by the "display name" of the module's owner, i.e. the person who is currently sharing the module data.
* The component should be fully previewable in Storybook, in two variants for both of the above possible circumstances.

## Implementation Brief
* Create `assets/js/components/ViewOnlyMenu/index.js` which exports the `ViewOnlyMenu` functional component.
* Similar to the implementation in `assets/js/components/UserMenu.js`,
	* the `Button` component will be used to render the `View only` text and icon and the `Menu` component will be used to render the dropdown menu.
	* Add a state variable `menuOpen` (bool) to determine when the menu is opened which will be passed on to the `menuOpen` prop of `Menu`.
	* Upon clicking the `Button` component, toggle the `menuOpen` to show/hide the menu.
	* Copy over the use of the `useClickAway` and `useKeyCodesInside` logic.
* Create `assets/js/components/ViewOnlyMenu/Description.js` which exports the `Description` functional component.
	* The component will render the first part of the dropdown menu, i.e the "View-only access lets you see ..." text along with the potential "Sign In With Google" button. 
	* Query the user data store via the `hasCapability` selector, passing `PERMISSION_AUTHENTICATE` to check if the user has the capability to authenticate with Site Kit.
	* If the user has the capability, render the text and "Sign In With Google" button else render the text with the "Learn More" link as per the designs in Figma.
	* The logic when clicking the "Sign In With Google" button should be similar to the `onButtonClick` function in `assets/js/components/setup/SetupUsingProxy.js` with the following notes:
		* The proxy setup URL can be obtained by querying the `core/site` datastore via the `proxySetupURL` selector.
		* Check if the site is connected by querying the `core/site` datastore via the `isConnected` selector.
		* Discard logic pertaining to `moduleReauthURL` and `connectAnalytics`.
		* Keep the `trackEvent` calls (`start_user_setup` and `start_site_setup`) but with the event category set to the current `viewContext` instead of `VIEW_CONTEXT_DASHBOARD_SPLASH`.
			* To get the context, use the `useViewContext` hook.
	* Note that the "Learn more" link should be part of a single translatable string and `createInterpolateElement`  is used to achieve that. The URL is https://sitekit.withgoogle.com/documentation/using-site-kit/dashboard-sharing/ (still TBC).
	* The component should return the markup for the above wrapped between a `li` tag, having the `mdc-list-item` class name.
	* The text can be wrapped within `p` tag.
* Create `assets/js/components/ViewOnlyMenu/SharedServices.js` which exports the `SharedServices` functional component.
	* The component will render the second part of the dropdown menu, i.e the "Shared Services" text along with the list of shared services.
	* To get the list of shared services, query the `getViewableModules` selector of the `core/user` data store which will return a list of module slugs.
	* Loop through the list of module slugs and render the `Service` (see below) component passing the module slug as prop.
	* The component should return the markup for the above wrapped between a `li` tag, having the `mdc-list-item` class name.
	* The "Shared services" can be wrapped within a `h4` tag and the list of shared services can be wrapped within a `ul` tag.
	* For the separator, a `li` tag can be used, with the `mdc-list-divider` class name.
		* `<li class="mdc-list-divider" role="separator"></li>`
* Create `assets/js/components/ViewOnlyMenu/Service.js` which exports the `Service` functional component.
	* It accepts a single prop `module` which is the module slug.
	* The component should render the service title, icon and owner.
	* To get the above data from the module's slug, query the `core/modules` data store via the `getModule` selector, passing the module slug as parameter.
	* The "Share by ..." text should be rendered only if the user has the capability to authenticate with Site Kit.
		* Query the user data store via the `hasCapability` selector, passing `PERMISSION_AUTHENTICATE` to check if the user has the capability to authenticate with Site Kit.
* Create `assets/js/components/ViewOnlyMenu/Tracking.js` which exports the `Tracking` functional component.
	* The component will render the third part of the dropdown menu, i.e the "Thanks for using Site Kit! ..." and the tracking checkbox.
	* The text can be wrapped within `p` tag.
	* Use `assets/js/components/OptIn.js` component to render the tracking checkbox and label.
	* The component should return the markup for the above wrapped between a `li` tag, having the `mdc-list-item` class name.
* Update `assets/js/components/ViewOnlyMenu/index.js` to include `SignIn`, `SharedServices` and `Tracking` functional components as children of the `Menu` component.
* Create `assets/js/components/ViewOnlyMenu/index.stories.js` which will contain the stories for `ViewOnlyMenu` for the 2 different versions of the component.
* All styles should go in `assets/sass/components/global/_googlesitekit-view-only-menu.scss`.

### Test Coverage
* No new tests to be added.

## QA Brief

- Verify the stories for `ViewOnlyMenu` display as expected and according to the designs (see notes below).
- Storybook for `develop`: https://google.github.io/site-kit-wp/storybook/develop/

![image](https://user-images.githubusercontent.com/18395600/163383270-ce75b178-bf6a-4e89-bcdf-e0c491872637.png)

<details>
<summary>
Story screenshots:
</summary>

**Can Authenticate:**

![image](https://user-images.githubusercontent.com/18395600/163392416-93f62f24-dc9e-4c10-be6d-bf88f8085925.png)

**Cannot Authenticate:**

![image](https://user-images.githubusercontent.com/18395600/163392910-6070f079-ef24-4d51-8659-7f2045a1297f.png)

</details>

**Notes:**

- The font family usage in the Figma design is a bit inconsistent with the general usage within Site Kit, I have therefore stuck to the default fonts for the markup elements specified in the IB. That is, Roboto for all the text barring the `h4` (the "Shared services" sub heading), which is Google Sans.
- The "Sign in with Google" button in Figma is also a bit inconsistent and I have stuck with the standard button design, with uppercase text.
- As the IB specified to reuse the `OptIn` component, the tracking checkbox & label is that provided by the `OptIn` component rather than what is shown on Figma. The `OptIn` component is what is shown in the Settings > Admin Settings > Tracking section.
![image](https://user-images.githubusercontent.com/18395600/163392568-3bfb9980-eefe-488d-8fbd-1e0b674a1918.png)
- Viewing the story at the "Small mobile" viewport can result in visible scrollbars, which is not realistic when considering a mobile browser. Instead, I'd recommend opening the story in a new tab and using the responsive devtools to view at this size. See screenshots below.

<details>
<summary>
Screenshots to illustrate:
</summary>

**Visible scrollbars in "Small mobile" Storybook viewport (320px wide):**

![image](https://user-images.githubusercontent.com/18395600/163392654-55289014-b484-47cb-97c7-473a38f8824b.png)

**Open in a new tab:**

![image](https://user-images.githubusercontent.com/18395600/163389496-5dcaf782-76f7-44e2-a7e0-509f0bd78d0e.png)

**Viewing in a new tab at 320px viewport:**

![image](https://user-images.githubusercontent.com/18395600/163392731-ae3bbe7c-c7f0-4eb6-a2ce-427296952b62.png)

## Changelog entry

* Add a dropdown menu for users using the "view-only" dashboard sharing mode.

**Comment by asvinb** [comment]
@felixarntz @aaemnnosttv A couple of questions:

- Where does "[Learn More](https://www.figma.com/file/i04x83uA9U8HHBANlhS7Db/Dashboard-Sharing?node-id=1512%3A17482)" link to?
- Is including the "Shared by..." [text in scope](https://www.figma.com/file/i04x83uA9U8HHBANlhS7Db/Dashboard-Sharing?node-id=1519%3A18181)?

**Comment by felixarntz** [comment]
@asvinb Great questions!
> Where does "[Learn More](https://www.figma.com/file/i04x83uA9U8HHBANlhS7Db/Dashboard-Sharing?node-id=1512%3A17482)" link to?

Let's put it to https://sitekit.withgoogle.com/documentation/using-site-kit/dashboard-sharing/ for now (doesn't exist yet), I'll open a follow-up issue to potentially update later based on where the real documentation goes.

> Is including the "Shared by..." [text in scope](https://www.figma.com/file/i04x83uA9U8HHBANlhS7Db/Dashboard-Sharing?node-id=1519%3A18181)?

Yes, good point. As indicated by the Figma files, the text should only be included if the user can authenticate. I've added an extra point with this to the ACs.

> Are we tracking any events?

Not for now, we'll define the events around dashboard sharing later, via a separate issue.

Regarding the current IB, that mostly looks good. Just 3 things:
* Per the second reply above, can you now incorporate the extra requirement around showing module owners to the users that can authenticate?
* For the note on `proxySetupURL`, let's extend that a bit. Basically, the button should work similar to the `onButtonClick` callback in the `SetupUsingProxy` component (only without the Analytics part of course since you can't connect that from the UI here). The most important extra part is that clicking the button should also fire the same `start_user_setup` / `start_site_setup` events under the same conditions, just the event category should be based on the current `viewContext` (see `useViewContext()` hook) instead of the hard-coded `VIEW_CONTEXT_DASHBOARD_SPLASH`.
* The `SignIn` name for the component is a bit odd to me, since the component only _optionally_ even includes a "Sign in" button, so I think we should base the name on the primary purpose of the component that even applies to users that cannot authenticate. Maybe simply `Description`?

**Comment by asvinb** [comment]
Thanks @felixarntz . IB has been updated as per your comments and agree with you, we can use `Description` for the top most component instead of `SignIn` 😁

**Comment by wpdarren** [comment]
### QA Update:  ❌ 

@techanvil just a few small styling difference which don't seem to be covered in your notes. 

- The Shared Services header font color is `#5F6368` in Figma but `#5b5b61` on Storybook
- The Service title font color is `#000000` in Figma but `#000000DE` on Storybook. 
- The divider color is `rgba(0, 0, 0, 0.2)` in Figma but `rgba(0,0,0,.12)` on Storybook. 

**Note:** we don't have the icon for external links in Figma, but we do on Storybook, which is what I would expect.

**Comment by techanvil** [comment]
> ### QA Update: 
> @techanvil just a few small styling difference which don't seem to be covered in your notes.
> 
> * The Shared Services header font color is `#5F6368` in Figma but `#5b5b61` on Storybook
> * The Service title font color is `#000000` in Figma but `#000000DE` on Storybook.
> * The divider color is `rgba(0, 0, 0, 0.2)` in Figma but `rgba(0,0,0,.12)` on Storybook.
> 
> **Note:** we don't have the icon for external links in Figma, but we do on Storybook, which is what I would expect.

Hi @wpdarren, thanks for pointing these out.

TLDR, I think these are fine. In a bit more detail:

- The Shared Services header is implemented as an `h4`, and the standard colour for an `h4` in our CSS is `#5b5b61` which is also defined as `$c-tertiary`. Looking at the use of colours within Site Kit it's more consistent to use this as a text colour as opposed to `#5F6368` which is in less common use and generally used for buttons & icons. So I think it makes sense to leave this as it is.
- The Service title font (in Storybook) uses the standard Material Design list item colour, we don't actually have any absolute black text that I am aware of, and this colour is consistent with that used in other menus e.g. the user menu and the date selector (although, strangely, not the help menu), I think this one is OK too.
- The divider colour (in Storybook) again is the standard one provided by Material Design, so I don't think we should change it here, given it's such a subtle difference anyway.

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

- The stories for `ViewOnlyMenu` display as expected and according to the figma designs, with the exception of the observations [noted here](https://github.com/google/site-kit-wp/issues/4812#issuecomment-1110944091) and in the QAB. 
- Tested on desktop and small screen sizes.

<details>
 <summary>Screenshots</summary>

![image](https://user-images.githubusercontent.com/73545194/165735958-0e7430d4-6640-4b28-a7a0-df07854aa6a9.png)
![image](https://user-images.githubusercontent.com/73545194/165735977-9ca71907-2c19-405f-8500-befe0063e2c4.png)
![image](https://user-images.githubusercontent.com/73545194/165735991-103a433b-d7bd-47c7-9f77-c70c7120ba51.png)
![image](https://user-images.githubusercontent.com/73545194/165736013-ef946fa2-6c82-4830-ac9c-92a07b045778.png)

</details>
